Option to keep deck playing on track load

mbacarella opened the pull request:
This is a proposed solution as discussed in issue #10548.

This is my first proposed contribution to this project and first to an open source C++ project. I apologize if I missed something important or did something dumb. Thanks for your consideration :)

ronso0 commented:
Thank you for working on this! And congratulations for your first contribution : )

Since this is your first contribution we need you to sign the [contributor agreement](https://docs.google.com/forms/d/e/1FAIpQLScC9QG327sjLL0eWftmfGUasxFFLxg6LCXJ2xHDYRzFIRqyiw/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ) and thereby allows Mixxx to include your changes.
Please comment here when you filled the form and let us know which name should appear in the contributor list.
> I apologize if I missed something important

Please install pre-commit as described in [CONTRIBUTING.md#git-workflow](https://github.com/mixxxdj/mixxx/blob/main/CONTRIBUTING.md#git-workflow) (I think [wiki/Using-Git#set-up-automatic-code-checking](https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking) is more comprehensive)

ronso0 commented:
> Please install pre-commit as described in [CONTRIBUTING.md#git-workflow](https://github.com/mixxxdj/mixxx/blob/main/CONTRIBUTING.md#git-workflow) (I think [wiki/Using-Git#set-up-automatic-code-checking](https://github.com/mixxxdj/mixxx/wiki/Using-Git#set-up-automatic-code-checking) is more comprehensive)

As long as no one started to review this PR you may as well incorporate the pre-commit fixes and sqash everything into one commit and rebase onto mixxx/main.

mbacarella commented:
> Please comment here when you filled the form and let us know which name should appear in the contributor list.

I filled out the form. `Michael Bacarella <m@bacarella.com>` is fine.

mbacarella committed:
Option to keep deck playing on track load (closes #10548)

mbacarella commented:
>As long as no one started to review this PR you may as well incorporate the pre-commit fixes and sqash everything into one commit and rebase onto mixxx/main.

I installed the pre-commit hooks, squashed, rebased and pushed. It complained about `qmlformat` but I couldn't figure out how to install it, so I skipped. I assume the server side may run it and let me know if there are any problems.

mbacarella commented:
I apologize for missing the instructions on installing Git pre-commit hooks the first go-round.

mbacarella committed:
Catch up to main (resolve merge conflict)

mbacarella committed:
Tabstops: remove old item from tabstop, add comboBoxLoadWhenDeckPlaying

mbacarella commented:
After the rebase, I noticed and fixed two more issues as separate commits and pushed. I thought the clang-format pre-commit hook would tell me to fix things but it passed for me locally without complaint. Not sure what to expect.

ronso0 commented on src/preferences/dialog/dlgprefdeckdlg.ui:
```diff
@@ -24,12 +24,12 @@
         </property>
         <property name="buddy">
-         <cstring>checkBoxDisallowLoadToPlayingDeck</cstring>
+         <cstring>comboBoxLoadWhenDeckPlaying</cstring>
```
I'm not sure why you changed this (changing translation texts requires new translation for every language).

Also, I thought the new option would a) load a track into a playing deck and b) play that right away.
Even though b) requires a), a) and b) are unrelated (I might want to load tracks into playing decks, but I don not want auto-play)

mbacarella replied:
It's actually deleting the checkbox entirely and replacing it with a combobox that offers more functionality:

1. load track into playing deck and play that right away
2. load track into playing deck but stop
3. don't allow loading into playing deck

See the discussion at the end of this issue? https://github.com/mixxxdj/mixxx/issues/10548#issuecomment-1268685304

ronso0 replied:
I should have taken a closer look, not just skimming through...

ronso0 commented:
If did `git rebase -i [mixxx repo]/main` [Catch up to main (resolve merge conflict)](https://github.com/mixxxdj/mixxx/pull/10944/commits/1c7940309ec3b5bceebcb4265a5356c29f99c9f6)
[1c79403](https://github.com/mixxxdj/mixxx/pull/10944/commits/1c7940309ec3b5bceebcb4265a5356c29f99c9f6) would not be necessary (you didn't touch legacyskinparser).

mbacarella commented:
> If did git rebase -i [mixxx repo]/main Catch up to main (resolve merge conflict) 1c79403 would not be necessary (you didn't touch legacyskinparser).

I can rebase this now, but I notice you have started a review.

ronso0 commented:
Even though I explicitely just added a comment it is marked as "reviewed", this is irritating.
Feel free to rebase!

daschuer commented on src/mixer/playermanager.cpp:
```diff
@@ -686,6 +686,26 @@ void PlayerManager::slotLoadLocationToPlayer(
 
+void PlayerManager::slotLoadLocationToPlayerMaybePlay(const QString& location, const QString& group) {
+    bool play = false;
+    LoadWhenDeckPlaying loadWhenDeckPlaying =
+            m_pConfig->exists(kConfigKeyLoadWhenDeckPlaying)
+        ? static_cast<LoadWhenDeckPlaying>(m_pConfig->getValueString(kConfigKeyLoadWhenDeckPlaying).toInt())
+        : kDefaultLoadWhenDeckPlaying;
```
```suggestion
    LoadWhenDeckPlaying loadWhenDeckPlaying =
            m_pConfig->getValue(kConfigKeyLoadWhenDeckPlaying, kDefaultLoadWhenDeckPlaying);
```

daschuer commented on src/mixer/playermanager.h:
```diff
@@ -188,9 +188,8 @@ class PlayerManager : public QObject, public PlayerManagerInterface {
     // Slots for loading tracks into a Player, which is either a Sampler or a Deck
     void slotLoadTrackToPlayer(TrackPointer pTrack, const QString& group, bool play = false);
     void slotLoadLocationToPlayer(const QString& location, const QString& group, bool play = false);
-    void slotLoadLocationToPlayerStopped(const QString& location, const QString& group) {
```
The default play in the two functions above should be probably removed to make it explicit.

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -178,6 +168,27 @@ DlgPrefDeck::DlgPrefDeck(QWidget* parent,
+    comboBoxLoadWhenDeckPlaying->addItem(tr("Reject"), static_cast<int>(LoadWhenDeckPlaying::Reject));
+    int loadWhenDeckPlaying;
+    if (m_pConfig->exists(kConfigKeyLoadWhenDeckPlaying)) {
+        loadWhenDeckPlaying = m_pConfig->getValueString(kConfigKeyLoadWhenDeckPlaying).toInt();
```
Do we need to use the default as well, if the value does not exists?

mbacarella replied:
Sorry, I don't follow. If it doesn't exist, won't it then get the default from the else-else?

daschuer replied:
Yes, right. 
It is just the issue that "m_pConfig->exists" is a redundant lookup of the value. It is better to use getValue() with a fallback value like purposed above

mbacarella replied:
Well yes but I'm trying to be mindful of not losing behavior previously configured by the user in a previous version. If the user disabled (or enabled) playing track protection I want to import that.

mbacarella replied:
I realize this is confusing because, in a different part of the code, I didn't use `getValue` with default when I should have, but I think in this case it makes sense to probe for existence of the new config variable first and fall back to loading the older config variable.

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -178,6 +168,27 @@ DlgPrefDeck::DlgPrefDeck(QWidget* parent,
+    connect(comboBoxLoadWhenDeckPlaying,
+            QOverload<int>::of(&QComboBox::currentIndexChanged),
+            this,
+            &DlgPrefDeck::slotSetLoadWhenDeckPlaying);
```
slotSetLoadWhenDeckPlaying()  fixes the relation of combobox index to value. Lets rename it to 
slotLoadWhenDeckPlayingIndexChanged and do the mapping in source code.

mbacarella replied:
Sorry, I don't follow this either.

I just did what the other combo boxes do (e.g. `slotSetTrackLoadMode`). Is there an example of your suggestion? :)

daschuer replied:
It looks like it is sufficient to just rename the function, because it is not a classic setter. It receives the index, fetches the value and than sets the the value. The current name does not expose that an index is expected.

ronso0 commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -178,6 +168,27 @@ DlgPrefDeck::DlgPrefDeck(QWidget* parent,
             this,
             &DlgPrefDeck::slotSetTrackLoadMode);
 
+    comboBoxLoadWhenDeckPlaying->addItem(tr("Allow"), static_cast<int>(LoadWhenDeckPlaying::Allow));
+    comboBoxLoadWhenDeckPlaying->addItem(tr("Allow, but stop deck"), static_cast<int>(LoadWhenDeckPlaying::AllowButStopDeck));
+    comboBoxLoadWhenDeckPlaying->addItem(tr("Reject"), static_cast<int>(LoadWhenDeckPlaying::Reject));
```
IMO it makes more sense to have the current default at the top (= the least invasive)
* Reject
* Allow, stop deck
* Allow, play from load point

mbacarella committed:
Incorporate code review feedback

mbacarella commented:
Did a bunch of the code review driven changes. *Didn't* rebase since a review actually started after @ronso0 said to rebase away :)

Anything else?

ronso0 commented:
Just a hint: usually code conversations (conversations pinned to a specific spot in the code) are marked as resolved by the reviewer. That makes it easier to check if/how the discussed changes were applied.

mbacarella commented:
> Just a hint: usually code conversations (conversations pinned to a specific spot in the code) are marked as resolved by the reviewer. That makes it easier to check if/how the discussed changes were applied.

My bad! Will bear that in mind for next time :)

mbacarella commented:
@ronso0 just to be clear I think this is ready to merge if there are no other blockers

(not sure what to do about the code style check; should I run clang-format on my code and push?)

ronso0 commented:
I only che.cked the UI until now, @daschuer has the final word re code here.

Update pre-commit maybe? Idk why it didn't complain about the indentations locally.
You may also download the generated patch and apply it manually. Click on Details next to the pre-commit check, on the top left click Summary, scroll down to Artifacts and grab [pre-commit.patch](https://github.com/mixxxdj/mixxx/suites/8719138046/artifacts/394105136)

daschuer reviewed (commented):
I have just tested this and it works good. Thank you. 
Just one comment for improvement and the pre-commit patch and we can merge this nice addition.

mbacarella committed:
Make AllowTrackLoadToPlayingDeck a const ConfigKey

mbacarella committed:
Run git-clang-format

mbacarella commented on src/util/dnd.cpp:
```diff
@@ -80,10 +81,24 @@ bool allowLoadToPlayer(
+    } else {
+        // support older version of this flag
+        allowLoadTrackIntoPlayingDeck =
+            pConfig->getValue<bool>(ConfigKey("[Controls]", "AllowTrackLoadToPlayingDeck"));
```
Done

mbacarella committed:
Actually try using the pre-commit.patch this time

daschuer reviewed (approved):
LGTM, Thank you.
